MdeModulePkg/Library/DxeCorePerformanceLib/DxeCorePerformanceLib.c

```diff
@@ -308,26 +319,37 @@ InternalGetSmmPerfData (
           SmmBootRecordData             = AllocateZeroPool (SmmBootRecordDataSize);
           SmmBootRecordDataRetrieved    = 0;
           ASSERT (SmmBootRecordData  != NULL);
-          SmmCommData->BootRecordData = (VOID *)((UINTN)SmmCommMemRegion->PhysicalStart + SMM_BOOT_RECORD_COMM_SIZE);
-          SmmCommData->BootRecordSize = ReservedMemSize - SMM_BOOT_RECORD_COMM_SIZE;
-          while (SmmBootRecordDataRetrieved < SmmBootRecordDataSize) {
-            Status = Communication->Communicate (Communication, SmmBootRecordCommBuffer, &CommSize);
-            ASSERT_EFI_ERROR (Status);
-            ASSERT_EFI_ERROR (SmmCommData->ReturnStatus);
-            if (SmmBootRecordDataRetrieved + SmmCommData->BootRecordSize > SmmBootRecordDataSize) {
-              CopyMem ((UINT8 *)SmmBootRecordData + SmmBootRecordDataRetrieved, SmmCommData->BootRecordData, SmmBootRecordDataSize - SmmBootRecordDataRetrieved);
-            } else {
-              CopyMem ((UINT8 *)SmmBootRecordData + SmmBootRecordDataRetrieved, SmmCommData->BootRecordData, SmmCommData->BootRecordSize);
+          // MU_CHANGE [BEGIN] - Standalone MM Perf Support
+          if (SmmBootRecordData != NULL) {
+            CurrentBootRecordData = (VOID *)((UINTN)SmmCommMemRegion->PhysicalStart + SMM_BOOT_RECORD_COMM_SIZE);
+            while (SmmBootRecordDataRetrieved < SmmBootRecordDataSize) {
+              // Note: Maximum comm buffer data payload size is ReservedMemSize - SMM_BOOT_RECORD_COMM_SIZE
+              BootRecordDataPayloadSize = MIN (
+                                            ReservedMemSize - SMM_BOOT_RECORD_COMM_SIZE,
+                                            SmmBootRecordDataSize - SmmBootRecordDataRetrieved
+                                            );
+              SmmCommBufferHeader->MessageLength = sizeof (SMM_BOOT_RECORD_COMMUNICATE) + BootRecordDataPayloadSize;
+              CommSize                           = sizeof (EFI_MM_COMMUNICATE_HEADER) + (UINTN)SmmCommBufferHeader->MessageLength;
+              Status                             = Communication->Communicate (Communication, SmmBootRecordCommBuffer, &CommSize);
+              ASSERT_EFI_ERROR (Status);
+              ASSERT_EFI_ERROR (SmmCommData->ReturnStatus);

```

Comment:
Do you want to finish the loop if these failures occur? Or better just return from here?

Reply:
I left an assert on the `Communicate()` status since but just returned on either an error from communicate itself or the handler to let the caller act on the handler status without an assert if they'd like.